File: go/dsa/algorithms/sliding-window/sliding-window-maximum.go
func maxSlidingWindow(nums []int, k int) []int {
    var q []int
    var res []int
    left := 0

    for right, n := range nums {
        // 1️⃣ Expand: Maintain monotonic property
        for len(q) > 0 && nums[q[len(q)-1]] <= n {
            q = q[:len(q)-1]
        }
        q = append(q, right)

        // 2️⃣ When window size reaches k
        if right - left + 1 == k {
            // 3️⃣ Record / Evaluate
            res = append(res, nums[q[0]])
        }
    }

    return res
}

File: cpp/dsa/algorithms/sliding-window/sliding-window-maximum.cpp
vector<int> maxSlidingWindow(vector<int>& nums, int k) {
    deque<int> dq;
    vector<int> res;
    int left = 0;

    for (int right = 0; right < nums.size(); right++) {
        // 1️⃣ Expand: Maintain monotonic property
        while (!dq.empty() && nums[dq.back()] <= nums[right]) {
            dq.pop_back();
        }
        dq.push_back(right);

        // 2️⃣ When window size reaches k
        if (right - left + 1 == k) {
            // 3️⃣ Record / Evaluate
            res.push_back(nums[dq.front()]);

            // 4️⃣ Slide window
            if (dq.front() == left) dq.pop_front();
            left++;
        }
    }
    return res;
}